algorithms/finetune/cal_ql.py

```diff
@@ -0,0 +1,1143 @@
+        group=config["group"],
+        name=config["name"],
+        id=str(uuid.uuid4()),
+    )
+    wandb.run.save()
+
+
+@torch.no_grad()
+def eval_actor(
+    env: gym.Env, actor: nn.Module, device: str, n_episodes: int, seed: int
+) -> np.ndarray:
+    env.seed(seed)
+    actor.eval()
+    episode_rewards = []
+    for _ in range(n_episodes):
+        state, done = env.reset(), False
+        episode_reward = 0.0
+        while not done:
+            action = actor.act(state, device)
+            state, reward, done, _ = env.step(action)
+            episode_reward += reward
+        episode_rewards.append(episode_reward)
+
+    actor.train()
+    return np.asarray(episode_rewards)
+
+
+def return_reward_range(dataset, max_episode_steps):
+    returns, lengths = [], []
+    ep_ret, ep_len = 0.0, 0
+    for r, d in zip(dataset["rewards"], dataset["terminals"]):
+        ep_ret += float(r)
+        ep_len += 1
+        if d or ep_len == max_episode_steps:
+            returns.append(ep_ret)
+            lengths.append(ep_len)
+            ep_ret, ep_len = 0.0, 0
+    lengths.append(ep_len)  # but still keep track of number of steps
+    assert sum(lengths) == len(dataset["rewards"])
+    return min(returns), max(returns)
+
+
+def get_return_to_go(dataset, gamma, max_episode_steps):
+    returns = []
+    ep_ret, ep_len = 0.0, 0
+    cur_rewards = []
+    terminals = []
+    for r, d in zip(dataset["rewards"], dataset["terminals"]):
+        ep_ret += float(r)
+        cur_rewards.append(float(r))
+        terminals.append(float(d))
+        ep_len += 1
+        if d or ep_len == max_episode_steps:
+            discounted_returns = [0] * ep_len
+            prev_return = 0
+            for i in range(ep_len):
+                discounted_returns[-i - 1] = cur_rewards[
+                    -i - 1
+                ] + gamma * prev_return * (1 - terminals[-i - 1])
+                prev_return = discounted_returns[-i - 1]
+            returns += discounted_returns
+            ep_ret, ep_len = 0.0, 0
+            cur_rewards = []
+            terminals = []
+    discounted_returns = [0] * ep_len
+    prev_return = 0
+    for i in range(ep_len):
+        discounted_returns[-i - 1] = cur_rewards[-i - 1] + gamma * prev_return * (
+            1 - terminals[-i - 1]
+        )
+        prev_return = discounted_returns[-i - 1]
+    returns += discounted_returns
+    return returns
+
+
+def modify_reward(dataset, env_name, max_episode_steps=1000):
+    if any(s in env_name for s in ("halfcheetah", "hopper", "walker2d")):
+        min_ret, max_ret = return_reward_range(dataset, max_episode_steps)
+        dataset["rewards"] /= max_ret - min_ret
+        dataset["rewards"] *= max_episode_steps
+        return {
+            "max_ret": max_ret,
+            "min_ret": min_ret,
+            "max_episode_steps": max_episode_steps,
+        }
+    elif "antmaze" in env_name:
+        dataset["rewards"] = dataset["rewards"] * 10.0 - 5.0
+    return {}
```

Comment:
why?

Reply:
We need to return additional information only in case of locomotion tasks

Reply:
kk, but still need typings